### DISCUSSION TITLE
How is Nextjs removing server-side top-level imports internally?

### DISCUSSION THREAD

**Discussion by hamza512b** [vercel/next.js#46606] [Help]
Hi! This question is not concern next.js per se, but I hope someone can answer me. I am building React SSR template, I have same feature as Next.js where you can get server side props and pass them to the client. I have problem that webpack is still importing the server-side top level imports in the client side. I wondering how did Next.js manage to fix this issue? Can you point me to where in source code this behavior is defined.

To put it differently, how does Next.js tree shake serverside code from client side?

**Comment by icyJoseph** [comment]
Next.js runs different builds. Three in fact, one for `nextRuntime: undefined`, which is used for the browser, `nextRuntime: "nodejs"`, and `nextRuntime: "edge"`.

For example, when running the node.js build, it imports GSSP, and the dependencies that uses.

Similarly, the client code bundle takes the default export of the page, and the dependencies that uses.

- https://next-code-elimination.vercel.app/

You can cause issues by defining variables and not using them in either client or server code, so it keeps it for both.


You can see the builds by defining a webpack function in your next.config.js file:

```js
webpack(config, {nextRuntime}) {
 console.log(nextRuntime)
 return config;
}
```
